sys/vfs/vfs.c

```diff
@@ -1104,22 +1104,49 @@ static inline int _fd_is_valid(int fd)
     return 0;
 }
 
+static bool _is_dir(vfs_mount_t *mountp, vfs_DIR *dir, const char *restrict path)
+{
+    const vfs_dir_ops_t *ops = mountp->fs->d_op;
+    if (!ops->opendir) {
+        return false;
+    }
+
+    int res = ops->opendir(dir, path, path);
+    if (res < 0) {
+        return false;
+    }
+
+    ops->closedir(dir);
+    return true;
+}
+
 int vfs_sysop_stat_from_fstat(vfs_mount_t *mountp, const char *restrict path, struct stat *restrict buf)
 {
     const vfs_file_ops_t * f_op = mountp->fs->f_op;
-    vfs_file_t opened = {
-        .mp = mountp,
-        /* As per definition of the `vfsfile_ops::open` field */
-        .f_op = f_op,
-        .private_data = { .ptr = NULL },
-        .pos = 0,
+
+    union {
+        vfs_file_t file;
+        vfs_DIR dir;
+    } filedir = {
+        .file = {
+            .mp = mountp,
+            /* As per definition of the `vfsfile_ops::open` field */
+            .f_op = f_op,
+            .private_data = { .ptr = NULL },
+            .pos = 0,
+        },
     };
```

Comment:
I think I understand the intention behind this union (to not have stack space for a vfs_file_t and a vfs_DIR allocated at the same time), but do we really need this optimization? It's not easy to read, and a compiler should (didn't test it) be smart enough to see that the stack space allocated for `opened` is dead by the time we call into `_is_dir` (and call over with the shorter stack frame), or it'd inline `_is_dir` anyway and see that it can alias a the unused opened file with an own `vfs_DIR` local to that area.

Reply:
> a compiler should (didn't test it) be smart enough 

A gave it a try 

<details><summary>this patch</summary>

```
diff --git a/sys/vfs/vfs.c b/sys/vfs/vfs.c
index 74ac84508f..7862fb069a 100644
--- a/sys/vfs/vfs.c
+++ b/sys/vfs/vfs.c
@@ -1127,29 +1127,25 @@ int vfs_sysop_stat_from_fstat(vfs_mount_t *mountp, const char *restrict path, st
 {
     const vfs_file_ops_t * f_op = mountp->fs->f_op;
 
-    union {
-        vfs_file_t file;
         vfs_DIR dir;
-    } filedir = {
-        .file = {
+        vfs_file_t file = {
             .mp = mountp,
             /* As per definition of the `vfsfile_ops::open` field */
             .f_op = f_op,
             .private_data = { .ptr = NULL },
             .pos = 0,
-        },
-    };
+       };
 
-    int err = f_op->open(&filedir.file, path, 0, 0, NULL);
+    int err = f_op->open(&file, path, 0, 0, NULL);
     if (err < 0) {
-        if (_is_dir(mountp, &filedir.dir, path)) {
+        if (_is_dir(mountp, &dir, path)) {
             buf->st_mode = S_IFDIR;
             return 0;
         }
         return err;
     }
-    err = f_op->fstat(&filedir.file, buf);
-    f_op->close(&filedir.file);
+    err = f_op->fstat(&file, buf);
+    f_op->close(&file);
     return err;
 }
```
</details>

And it's unfortunately not smart enough:

#### union
```
> ls /nvm0
test.txt	30 B
hello/
main.c	886 B
total 2 files

> ps
	pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) ( 8193) |  0x806eaa0 |  0x806eaa0
	  1 | idle                 | pending  Q |  15 |   8192 (  436) ( 7756) |  0x80673a0 |  0x8069200 
	  2 | main                 | running  Q |   7 |  12288 ( 3452) ( 8836) |  0x80693a0 |  0x806c200 
	    | SUM                  |            |     |  28672 ( 3888) (24784)
```

#### individual variables

```
> ls /nvm0
test.txt	30 B
hello/
main.c	886 B
total 2 files

> ps
	pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) ( 8193) |  0x806eaa0 |  0x806eaa0
	  1 | idle                 | pending  Q |  15 |   8192 (  436) ( 7756) |  0x80673a0 |  0x8069200 
	  2 | main                 | running  Q |   7 |  12288 ( 3548) ( 8740) |  0x80693a0 |  0x806c200 
	    | SUM                  |            |     |  28672 ( 3984) (24688)
```

 \+ 96 bytes on the `main` stack with `CFLAGS += -DFATFS_FFCONF_OPT_FS_EXFAT=1`

Reply:
As much as I'd like to understand what is going wrong for the compiler to be smarter here, I don't have the time to dig down on it -- thanks for doing the test.